src/hooks/useSearchPusherUpdates.ts

```diff
@@ -0,0 +1,75 @@
+import {useEffect} from 'react';
+import type {OnyxCollection} from 'react-native-onyx';
+import type {SearchQueryJSON} from '@components/Search/types';
+import {search} from '@libs/actions/Search';
+import CONST from '@src/CONST';
+import type {ReportActions, Transaction} from '@src/types/onyx';
+import usePrevious from './usePrevious';
+
+/**
+ * Hook to detect new transactions or report actions from Pusher and trigger a search refresh
+ */
+function useSearchPusherUpdates({
+    isOffline,
+    queryJSON,
+    transactions,
+    reportActions,
+}: {
+    isOffline: boolean;
+    queryJSON: SearchQueryJSON;
+    transactions?: OnyxCollection<Transaction>;
+    reportActions?: OnyxCollection<ReportActions>;
+}) {
+    const previousTransactions = usePrevious(transactions);
+    const previousReportActions = usePrevious(reportActions);
+    const isChat = queryJSON.type === CONST.SEARCH.DATA_TYPES.CHAT;
+
+    // Detect Pusher updates and trigger search
+    useEffect(() => {
+        if (isOffline) {
+            return;
+        }
+
+        // Only proceed if we have stable references to compare
+        if (isChat && (!reportActions || !previousReportActions)) {
+            return;
+        }
+
+        if (!isChat && (!transactions || !previousTransactions)) {
+            return;
+        }
+
+        // For chat, check if there are new report actions
+        if (isChat) {
+            const currentReportActionsIDs = Object.values(reportActions ?? {})
+                .map((actions) => Object.keys(actions ?? {}))
+                .flat();
+
+            const previousReportActionsIDs = Object.values(previousReportActions ?? {})
+                .map((actions) => Object.keys(actions ?? {}))
+                .flat();
+
+            // Only trigger search if Pusher added new report actions
+            const hasNewReportActions = currentReportActionsIDs.length > previousReportActionsIDs.length && currentReportActionsIDs.some((id) => !previousReportActionsIDs.includes(id));
```

Comment:
Shall we use Set for `previousReportActionsIDs`? Improvement can be significant when there are many report actions.

```suggestion
            const hasNewReportActions = currentReportActionsIDs.length > previousReportActionsIDs.length && currentReportActionsIDs.some((id) => !previousReportActionsIDs.has(id));
```

Reply:
Yep, good suggestion. Changing it, thanks!